### PR TITLE
/catalog/component pages

### PR DESCRIPTION
**Pull request by nottrobin**
QA
--

Check links on `/hardware` pages and `/catalog` section work as well as it does on live:

- [ ] Go to e.g. https://certification-ubuntu-com-canonical-web-and-design-pr-42.run.demo.haus/hardware/201612-25299, check all the links work
- [ ] See that a URL like https://certification-ubuntu-com-canonical-web-and-design-pr-42.run.demo.haus/catalog/component/10ec:8168 works as expected
- [ ] Check pagination work
- [ ] Check more complex URLs like https://certification-ubuntu-com-canonical-web-and-design-pr-42.run.demo.haus/catalog/component/pci/8086%3Aa170/ and https://certification-ubuntu-com-canonical-web-and-design-pr-42.run.demo.haus/catalog/component/dmi/4755/dmi%3AUnknown/ redirect to the simpler form, https://certification-ubuntu-com-canonical-web-and-design-pr-42.run.demo.haus/catalog/component/dmi%3AUnknown, and that that page works okay

### CODE REVIEW

File: webapp/app.py [hunk 1]
```diff
@@ -461,3 +461,36 @@ def component_details(id):
             machines, key=lambda machine: machine["canonical_id"], reverse=True
         ),
     )
+
+
+@app.route("/catalog/component/<identifier>")
+def catalog_component(identifier):
```
Comment (codersquid) [inline webapp/app.py]: I propose an API change to simplify this. Here are my first thoughts.

/api/v1/certifiedmodeldevice/{identifier}
it returns a list of certifiedmodeldevicess dicts that contain a list of certifiedmodels

We can work out the required filters.

File: webapp/app.py [hunk 2]
```diff
@@ -461,3 +461,36 @@ def component_details(id):
...
+
+    if not devices:
+        flask.abort(404)
+
+    canonical_ids = [device["canonical_id"] for device in devices]
+
+    # Only get the first 267 canonical_ids, as the URL will be too long
+    # otherwise
+    models_response = api.certifiedmodels(
+        canonical_id__in=",".join(canonical_ids[:267]),
```
Comment (codersquid) [inline webapp/app.py:483]: This is one of the things that would get obviated by the above api proposal.

File: .env [hunk 3]
```diff
@@ -1,2 +1,3 @@
 PORT=8034
 FLASK_DEBUG=true
+DEVEL=true
```
Comment (codersquid) [inline .env:3]: Is this an accidental checkin? Or when you deploy does .env get replaced?

File: redirects.yaml [hunk 4]
```diff
@@ -1,3 +1,4 @@
 components/hardware/(?P<canonical_id>[^/]+)/?: /components?canonical_id={canonical_id}
 certification: /
 certification/(?P<path>.+): /{path}
+catalog/component/[^/]+/([^/]+/)?(?P<identifier>[^/]+)/?: /catalog/component/{identifier}
```
Comment (codersquid) [inline redirects.yaml]: re discussion in irc, we need additional redirects. one for subsystem identifier.

we're still discussing requirements for the others.

File: webapp/app.py (hunk 1 above)
Comment (nottrobin) [reply webapp/app.py]: That looks great. But could we merge it this way for now?

File: webapp/app.py (hunk 2 above)
Comment (nottrobin) [reply webapp/app.py:483]: Indeed =)

File: .env (hunk 3 above)
Comment (nottrobin) [reply .env:3]: These env settings are for local development. `DEVEL` makes the server STDOUT output appear in shiny colours.

File: redirects.yaml (hunk 4 above)
Comment (nottrobin) [reply redirects.yaml]: I believe this is done.

**Comment by codersquid** [comment]
I'm seeing a different set of results for the video card for this machine, as an example.

https://certification.ubuntu.com/hardware/201003-5447/
https://certification-ubuntu-com-canonical-web-and-design-pr-42.run.demo.haus/hardware/201003-5447

and I need to work out whether it is an api thing, a frontend thing, or both.

**Comment by nottrobin** [comment]
@codersquid here's the logic of what my page is doing: it queries https://certification.canonical.com/api/v1/certifiedmodeldevices/?format=json&identifier=102b:0532&subsystem=1028:028c (1 result) then uses that canonical_id to query https://certification.canonical.com/api/v1/certifiedmodels/?format=json&canonical_id=201003-5447 = 1 result. Is that wrong?

**Comment by codersquid** [comment]
Nope, that looks good. I'm approving this. We might get a request to include the vendor_id as a filter, I'll open an issue if that happens.

**Review by codersquid — APPROVED** [review]

**Pull request merged**